File: src/open/elevationData.ts
import { Pbf as Protobuf } from 's2-tools';
import { deltaDecodeArray, deltaEncodeArray } from '../util';

import type { Extents } from '.';

/** Elevation object to read from */
export class ElevationData {
  data: number[] = [];
  extent: Extents = 8_192;
  size: number = 0;
  min: number = 0;
  max: number = 0;

  /**
   * @param tag - the tag to read
   * @param elevationData - the elevation data to mutate
   * @param pbf - the Protobuf to pull the appropriate data from
   * @internal
   */
  _read(tag: number, elevationData: ElevationData, pbf: Protobuf): void {
    if (tag === 0) elevationData.extent = pbf.readVarint() as Extents;
    else if (tag === 1) elevationData.size = pbf.readVarint();
    else if (tag === 2) elevationData.min = pbf.readFloat();
    else if (tag === 3) elevationData.max = pbf.readFloat();
    else if (tag === 4) {
      elevationData.data = deltaDecodeArray(pbf.readPackedVarint()).map((v) =>
        unmapValue(v, elevationData.min, elevationData.max, elevationData.extent),
      );
    }
  }
}

/**
 * Elevation data input. It is assumed data is the actual elevation data of each point in the tile
 */
export interface ElevationInput {
  /** The size of the square tile. The length and width are the same size */
  size: number;
  /**
   * The actual elevation data assumed to be floating point (32-bit) precise.
   * It is assumed to find data you can use: `index = y * size + x`
   */
  data: number[];
  /** The extent defines the 0->extent range to remap the data to  */
  extent: Extents;
}

/**
 * @param input - the data to encode
 * @returns - the encoded data
 */
export function writeElevationData(input: ElevationInput): Uint8Array {
  const pbf = new Protobuf();

  const { data, size, extent } = input;
  const max = data.reduce((a, b) => Math.max(a, b), 0);
  const min = data.reduce((a, b) => Math.min(a, b), 0);
  const re_mapped = data.map((v) => remapValue(v, min, max, extent));
  const d_coded = deltaEncodeArray(re_mapped);

  pbf.writeVarintField(0, extent);
  pbf.writeVarintField(1, size);
  pbf.writeFloatField(2, min);
  pbf.writeFloatField(3, max);
  pbf.writePackedVarint(4, d_coded);

  return pbf.commit();
}

/**
 * @param value - input value to remap
 * @param min - min value
 * @param max - max value
 * @param extent - extent defines the 0->extent range to remap to
 * @returns - remapped value
 */
function remapValue(value: number, min: number, max: number, extent: number): number {
  return Math.round(((value - min) * extent) / (max - min));
}

/**
 * @param value - input value to unmap
 * @param min - min value
 * @param max - max value
 * @param extent - extent defines the 0->extent range to remap from
 * @returns - unremapped value
 */
function unmapValue(value: number, min: number, max: number, extent: number): number {
  return (value * (max - min)) / extent + min;
}

/**
 * @param r - red
 * @param g - green
 * @param b - blue
 * @returns - elevation
 */
export function convertTerrariumElevationData(r: number, g: number, b: number): number {
  return r * 256.0 + g + b / 256.0 - 32768.0;
}

/**
 * @param r - red
 * @param g - green
 * @param b - blue
 * @returns - elevation
 */
export function convertMapboxElevationData(r: number, g: number, b: number): number {
  return -10000 + (r * 256 * 256 + g * 256 + b) * 0.1;
}

File: rust/open/elevation_data.rs
use alloc::vec::Vec;

use crate::{delta_decode_array, delta_encode_array, open::Extent};

use libm::round;
use pbf::{ProtoRead, ProtoWrite, Protobuf};

/// Elevation object to read from
#[derive(Default, Debug)]
pub struct ElevationData {
    data: Vec<f64>,
    extent: Extent,
    size: f64,
    min: f64,
    max: f64,
}
impl ProtoRead for ElevationData {
    fn read(&mut self, tag: u64, pb: &mut Protobuf) {
        match tag {
            0 => self.extent = pb.read_varint(),
            1 => self.size = pb.read_varint(),
            2 => self.min = pb.read_varint(),
            3 => self.max = pb.read_varint(),
            4 => {
                self.data = delta_decode_array(&pb.read_packed())
                    .into_iter()
                    .map(|v| unmap_value(v as f64, self.min, self.max, self.extent.into()))
                    .collect()
            }
            _ => panic!("unknown tag: {}", tag),
        }
    }
}
impl ProtoWrite for ElevationData {
    fn write(&self, pb: &mut Protobuf) {
        let max = self.data.iter().fold(0.0, |a, b| f64::max(a, *b));
        let min = self.data.iter().fold(0.0, |a, b| f64::min(a, *b));
        let re_mapped: Vec<u32> =
            self.data.iter().map(|v| remap_value(*v, min, max, self.extent.into())).collect();
        let d_coded = delta_encode_array(&re_mapped);

        pb.write_varint_field(0, self.extent);
        pb.write_varint_field(1, self.size);
        pb.write_varint_field(2, min);
        pb.write_varint_field(3, max);
        pb.write_packed_varint(4, &d_coded);
    }
}

/// map the value to the range 0->extent
fn remap_value(value: f64, min: f64, max: f64, extent: f64) -> u32 {
    round(((value - min) * extent) / (max - min)) as u32
}

/// map the value back to floats
fn unmap_value(value: f64, min: f64, max: f64, extent: f64) -> f64 {
    (value * (max - min)) / extent + min
}

/// convert rgb to elevation using terrarium formula
pub fn convert_terrarium_elevation_data(r: u8, g: u8, b: u8) -> f64 {
    r as f64 * 256. + g as f64 + b as f64 / 256. - 32768.
}

/// convert rgb to elevation using mapbox formula
pub fn convert_mapbox_elevation_data(r: u8, g: u8, b: u8) -> f64 {
    -10000. + (r as f64 * 256. * 256. + g as f64 * 256. + b as f64) * 0.1
}
